qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -0,0 +1,118 @@
+
+        Args:
+            linear (list[list] or ndarray or QuantumCircuit):
+                either an n x n matrix, describing the linear function,
+                or a quantum circuit composed of linear gates only
+                (currently supported gates are CX and SWAP).
+
+            validate_input: if True, performs more expensive input validation checks,
+                such as checking that a given n x n matrix is invertible.
+
+        Raises:
+            CircuitError: if the input is invalid:
+                either a matrix is non {square, binary, invertible},
+                or a quantum circuit contains non-linear gates.
+        """
+        if not isinstance(linear, (list, np.ndarray, QuantumCircuit)):
+            raise CircuitError(
+                "A linear function must be represented either by a list, "
+                "a numpy array, or a quantum circuit with linear gates."
+            )
+
+        if isinstance(linear, QuantumCircuit):
+            # The following function will raise a CircuitError if there are nonlinear gates.
+            self._mat = _linear_quantum_circuit_to_mat(linear)
+
+        else:
+            # Check that the matrix is square
+            if not all(len(row) == len(linear) for row in linear):
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Check that all entries are either 0 or 1
+            for row in linear:
+                for entry in row:
+                    if entry not in (0, 1):
+                        raise CircuitError("A linear function must only contain 0s or 1s.")
+
+            # Optionally, check that the matrix is invertible
+            if validate_input:
+                det = np.linalg.det(linear) % 2
+                if not np.allclose(det, 1):
+                    raise CircuitError(
+                        "A linear function must be represented by an invertible matrix."
+                    )
+
+            self._mat = linear
+
+        super().__init__(name="linear_function", num_qubits=len(self._mat), params=[])
+
+    def _define(self):
+        """Populates self.definition with a decomposition of this gate."""
+        self.definition = self.synthesize()
+
+    def synthesize(self):
+        """Synthesize ``qiskit.circuit.library.LinearFunction``.
+
+        Returns:
+            QuantumCircuit: A circuit implementing the evolution.
+        """
+        from qiskit.transpiler.synthesis import cnot_synth
+
+        return cnot_synth(self._mat)
+
+    @property
+    def matrix(self):
+        """Returns the n x n matrix representing this linear function"""
+        return self._mat
+
+
+def _linear_quantum_circuit_to_mat(qc: QuantumCircuit):
+    """This creates a n x n matrix corresponding to the given linear quantum circuit."""
+    nq = qc.num_qubits
+    mat = np.eye(nq, nq, dtype=bool)
+    bit_indices = {bit: idx for idx, bit in enumerate(qc.qubits)}
```

Comment:
you don't really need `bit_indices` - `qc.find_bit(qubit).index` will do the job just as efficiently

Reply:
Thanks! Changed the code to use `find_bit`.